File: compare-sort-algorithms/src/radix_sort.go
package main

import (
	"bufio"
	"flag"
	"fmt"
	"log"
	"math"
	"os"
	"path/filepath"
	"strconv"
	"strings"
	"time"
)

// getMax finds the maximum value in the array
func getMax(arr []int) int {
	max := arr[0]
	for i := 1; i < len(arr); i++ {
		if arr[i] > max {
			max = arr[i]
		}
	}
	return max
}

// getMin finds the minimum value in the array
func getMin(arr []int) int {
	min := arr[0]
	for i := 1; i < len(arr); i++ {
		if arr[i] < min {
			min = arr[i]
		}
	}
	return min
}

// countingSortByDigit sorts the array based on the digit at the given position
func countingSortByDigit(arr []int, exp int, offset int) {
	n := len(arr)
	output := make([]int, n)
	count := make([]int, 10)

	// Store count of occurrences in count[]
	for i := 0; i < n; i++ {
		// Add offset to handle negative numbers
		adjustedValue := arr[i] + offset
		digit := (adjustedValue / exp) % 10
		count[digit]++
	}

	// Change count[i] so that count[i] now contains
	// actual position of this digit in output[]
	for i := 1; i < 10; i++ {
		count[i] += count[i-1]
	}

	// Build the output array
	for i := n - 1; i >= 0; i-- {
		adjustedValue := arr[i] + offset
		digit := (adjustedValue / exp) % 10
		output[count[digit]-1] = arr[i]
		count[digit]--
	}

	// Copy the output array to arr[]
	for i := 0; i < n; i++ {
		arr[i] = output[i]
	}
}

// radixSort implements the radix sort algorithm
func radixSort(arr []int) {
	if len(arr) <= 1 {
		return
	}

	// Find the minimum value for handling negative numbers
	min := getMin(arr)
	offset := 0
	if min < 0 {
		offset = -min
	}

	// Find the maximum value to know number of digits
	max := getMax(arr)
	if min < 0 {
		// When we have negative numbers, we need to consider the absolute values
		max = int(math.Max(float64(max), float64(abs(min))))
	}
	max += offset // Adjust max for the offset

	// Do counting sort for every digit
	for exp := 1; max/exp > 0; exp *= 10 {
		countingSortByDigit(arr, exp, offset)
	}

	// Remove the offset if any was applied
	if offset > 0 {
		for i := 0; i < len(arr); i++ {
			arr[i] -= 0 // No need to adjust since we're sorting in-place and maintaining relative order
		}
	}
}

// abs returns the absolute value of an integer
func abs(x int) int {
	if x < 0 {
		return -x
	}
	return x
}

// readNumbersFromFile reads integers from a file and returns them as a slice
func readNumbersFromFile(filename string) ([]int, error) {
	file, err := os.Open(filename)
	if err != nil {
		return nil, err
	}
	defer file.Close()

	var numbers []int
	scanner := bufio.NewScanner(file)
	for scanner.Scan() {
		line := strings.TrimSpace(scanner.Text())
		if line == "" {
			continue
		}

		num, err := strconv.Atoi(line)
		if err != nil {
			// Skip lines that cannot be converted to integers
			continue
		}

		numbers = append(numbers, num)
	}

	if err := scanner.Err(); err != nil {
		return nil, err
	}

	return numbers, nil
}

// isSorted checks if the array is sorted
func isSorted(arr []int) bool {
	for i := 0; i < len(arr)-1; i++ {
		if arr[i] > arr[i+1] {
			return false
		}
	}
	return true
}

func main() {
	fmt.Println("Go Radix Sort Performance Test")
	fmt.Println("==============================")

	// Parse command-line flags
	filePtr := flag.String("file", "random_list.txt", "Path to input file with numbers")
	outputPtr := flag.String("output", "results_go.txt", "Path to output file for results")
	flag.Parse()

	fmt.Printf("Using dataset: %s\n", *filePtr)

	// Read data from file
	fmt.Println("Reading data from file...")
	data, err := readNumbersFromFile(*filePtr)
	if err != nil {
		log.Fatalf("Error reading file: %v", err)
	}
	fmt.Printf("Data size: %d integers\n", len(data))

	// Create a copy for sorting (to preserve original)
	dataCopy := make([]int, len(data))
	copy(dataCopy, data)

	// Measure sorting time
	fmt.Println("Starting Radix Sort...")
	startTime := time.Now()
	radixSort(dataCopy)
	executionTime := time.Since(startTime).Seconds()

	// Verify the array is sorted
	sorted := isSorted(dataCopy)
	sortedStr := "SUCCESS"
	if !sorted {
		sortedStr = "FAILED"
	}

	// Results
	fmt.Printf("Sorting completed: %s\n", sortedStr)
	fmt.Printf("Execution time: %.6f seconds\n", executionTime)
	fmt.Printf("Elements per second: %.0f\n", float64(len(data))/executionTime)

	// Create output directory if it doesn't exist
	outputDir := filepath.Dir(*outputPtr)
	if outputDir != "." && outputDir != "" {
		err = os.MkdirAll(outputDir, 0755)
		if err != nil {
			log.Fatalf("Error creating output directory: %v", err)
		}
	}

	// Save results to file
	resultsFile, err := os.Create(*outputPtr)
	if err != nil {
		log.Fatalf("Error creating results file: %v", err)
	}
	defer resultsFile.Close()

	fmt.Fprintf(resultsFile, "Go Radix Sort Results\n")
	fmt.Fprintf(resultsFile, "Data size: %d\n", len(data))
	fmt.Fprintf(resultsFile, "Execution time: %.6f seconds\n", executionTime)
	fmt.Fprintf(resultsFile, "Elements per second: %.0f\n", float64(len(data))/executionTime)
	fmt.Fprintf(resultsFile, "Sorted correctly: %t\n", sorted)

	fmt.Printf("Results saved to %s\n", *outputPtr)
}

File: compare-sort-algorithms/src/radix_sort.cpp
#include <iostream>
#include <vector>
#include <fstream>
#include <chrono>
#include <iomanip>
#include <algorithm>

class RadixSort {
private:
    // A function to get maximum value in arr[]
    static int getMax(const std::vector<int>& arr) {
        int max = arr[0];
        for (size_t i = 1; i < arr.size(); i++) {
            if (arr[i] > max) {
                max = arr[i];
            }
        }
        return max;
    }
    
    // A function to get minimum value in arr[] (for handling negatives)
    static int getMin(const std::vector<int>& arr) {
        int min = arr[0];
        for (size_t i = 1; i < arr.size(); i++) {
            if (arr[i] < min) {
                min = arr[i];
            }
        }
        return min;
    }
    
    // A function to do counting sort based on the digit
    static void countSort(std::vector<int>& arr, int exp, int offset = 0) {
        std::vector<int> output(arr.size());
        int count[10] = {0};
        
        // Store count of occurrences in count[]
        for (size_t i = 0; i < arr.size(); i++) {
            int adjusted = arr[i] + offset;
            count[(adjusted / exp) % 10]++;
        }
        
        // Change count[i] so that count[i] contains actual
        // position of this digit in output[]
        for (int i = 1; i < 10; i++) {
            count[i] += count[i - 1];
        }
        
        // Build the output array
        for (int i = arr.size() - 1; i >= 0; i--) {
            int adjusted = arr[i] + offset;
            output[count[(adjusted / exp) % 10] - 1] = arr[i];
            count[(adjusted / exp) % 10]--;
        }
        
        // Copy the output array to arr[]
        for (size_t i = 0; i < arr.size(); i++) {
            arr[i] = output[i];
        }
    }
    
public:
    static void sort(std::vector<int>& arr) {
        if (arr.empty()) {
            return;
        }
        
        // To handle negative numbers, find the minimum value and offset all values to make them positive
        int min = getMin(arr);
        int offset = (min < 0) ? -min : 0; // If min is negative, offset by its absolute value
        
        // Find the maximum number to know number of digits
        int max = getMax(arr);
        max += offset; // Adjust max for the offset
        
        // Do counting sort for every digit
        for (int exp = 1; max / exp > 0; exp *= 10) {
            countSort(arr, exp, offset);
        }
        
        if (offset > 0) {
            // Remove the offset we added
            for (size_t i = 0; i < arr.size(); i++) {
                arr[i] -= 0; // No need to adjust since we're sorting in-place and maintaining relative order
            }
        }
    }
};

std::vector<int> readNumbersFromFile(const std::string& filename = "random_list.txt") {
    std::vector<int> numbers;
    std::ifstream file(filename);
    
    if (!file.is_open()) {
        std::cerr << "Error: " << filename << " not found. Please run generate_data.py first." << std::endl;
        exit(1);
    }
    
    int number;
    while (file >> number) {
        numbers.push_back(number);
    }
    
    file.close();
    return numbers;
}

bool isSorted(const std::vector<int>& arr) {
    for (size_t i = 0; i < arr.size() - 1; i++) {
        if (arr[i] > arr[i + 1]) {
            return false;
        }
    }
    return true;
}

int main(int argc, char* argv[]) {
    std::cout << "C++ Radix Sort Performance Test" << std::endl;
    std::cout << "===============================" << std::endl;
    
    // Get input and output filenames from command line arguments
    std::string filename = "random_list.txt";
    std::string resultFilename = "results_cpp.txt";
    
    if (argc > 1) {
        filename = argv[1];
        std::cout << "Using dataset: " << filename << std::endl;
    }
    
    if (argc > 2) {
        resultFilename = argv[2];
    }
    
    // Read data from file
    std::cout << "Reading data from file..." << std::endl;
    std::vector<int> data = readNumbersFromFile(filename);
    std::cout << "Data size: " << data.size() << " integers" << std::endl;
    
    // Create a copy for sorting (to preserve original)
    std::vector<int> dataCopy = data;
    
    // Measure sorting time
    std::cout << "Starting Radix Sort..." << std::endl;
    auto startTime = std::chrono::high_resolution_clock::now();
    RadixSort::sort(dataCopy);
    auto endTime = std::chrono::high_resolution_clock::now();
    
    auto duration = std::chrono::duration_cast<std::chrono::microseconds>(endTime - startTime);
    double executionTime = duration.count() / 1000000.0; // Convert to seconds
    
    // Verify the array is sorted
    bool sorted = isSorted(dataCopy);
    
    // Results
    std::cout << "Sorting completed: " << (sorted ? "SUCCESS" : "FAILED") << std::endl;
    std::cout << std::fixed << std::setprecision(6);
    std::cout << "Execution time: " << executionTime << " seconds" << std::endl;
    std::cout << std::fixed << std::setprecision(0);
    std::cout << "Elements per second: " << data.size() / executionTime << std::endl;
    
    // Output filename was already defined above from command line args
    
    std::ofstream resultFile(resultFilename);
    resultFile << "C++ Radix Sort Results" << std::endl;
    resultFile << "Data size: " << data.size() << std::endl;
    resultFile << std::fixed << std::setprecision(6);
    resultFile << "Execution time: " << executionTime << " seconds" << std::endl;
    resultFile << std::fixed << std::setprecision(0);
    resultFile << "Elements per second: " << data.size() / executionTime << std::endl;
    resultFile << "Sorted correctly: " << (sorted ? "true" : "false") << std::endl;
    resultFile.close();
    
    std::cout << "Results saved to " << resultFilename << std::endl;
    
    return 0;
}
